src/index.js

```diff
@@ -0,0 +1,329 @@
+    }
+  }
+
+  /**
+   * Wrapper function around #connectProducer where it first checks whether the producer has been connected previously
+   * @private
+   */
+  async #initProducer() {
+    try {
+      if (!this.#isProducerConnected) {
+        console.warn('Kafka producer is not connected. Retrying...');
+        await this.#connectProducer();
+      }
+    } catch (error) {
+      throw new Error(`Error initializing producer: ${error}`);
+    }
+  }
+
+  /**
+   * Wrapper function around #connectConsumer where it first checks whether the consumer has been connected previously
+   * @private
+   */
+  async #initConsumer() {
+    try {
+      if (!this.#isConsumerConnected) {
+        console.warn('Kafka consumer is not connected. Retrying...');
+        await this.#connectConsumer();
+      }
+    } catch (error) {
+      throw new Error(`Error initializing producer: ${error}`);
+    }
+  }
+
+  /**
+   * Sends an encoded message to a topic. Encodes the message data using this.#registry.encode
+   * @param {String} topic The topic to send the message to
+   * @param {Object} message The message to send to a topic
+   * @public
+   */
+  async sendMessage(topic, message) {
+    try {
+      await this.#initProducer();
+
+      if (this.#isProducerConnected) {
+        const subject = `${topic}-value`;
+        const id = await this.#registry.getRegistryId(subject, 'latest');
+
+        console.log(`Using schema ${topic}-value@latest (id: ${id})`);
+
+        const encodedMessage = await this.#registry.encode(id, message);
+
+        this.#producer.produce(
+          topic,
+          null, // Partition, null for automatic partitioning
+          Buffer.from(encodedMessage),
+          `${topic}-schema`, // Key
+        );
+
+        console.log(`Successfully published data to topic: ${topic}`);
+      } else {
+        console.error('Major issue with the kafka producer init process.');
+        throw new Error('Unable to initialize kafka producer');
+      }
+    } catch (error) {
+      console.error(
+        `Error occurred in sending message to topic ${topic}: ${error}`,
+      );
+      throw new Error(
+        `Error occurred in sending message to topic ${topic}: ${error}`,
+      );
+    }
+  }
+
+  /**
+   * Consumes a message from a topic. Decodes the message using this.#registry.decode
+   * @param {String} topic The topic to consume the message from
+   * @callback onMessage
+   * @param {onMessage} onMessage A function that processes the decoded message data received by consumer
+   * @public
+   */
+  async consumeMessage(topic, onMessage) {
+    try {
+      await this.#initConsumer();
+
+      if (this.#isConsumerConnected) {
+        this.#consumer.subscribe([topic]);
+        console.log(`Subscribed to topic ${topic}`);
+
+        if (!this.#intervalId) {
+          this.#intervalId = setInterval(function () {
+            this.#consumer.consume(10); // Read 10 messages every 1000 milliseconds.
+          }, 1000);
```

Comment:
BINDING issue here with this.

```suggestion
          this.#intervalId = setInterval(() => {
        this.#consumer.consume(10);
      }, 1000);
```

Reply:
Binding issue is resolved in [commit](https://github.com/cinemataztic/big-evil-kafka/pull/1/commits/6186bf34638fb4201f5d0c2563132e8306f42b08)